### PR TITLE
Allow simulator quantum state representations to be used in qis

### LINKED ISSUE #4582: Consolidate quantum state representations in cirq
**Is your design idea/issue related to a use case or problem? Please describe.**

There is a zoo of quantum states and special representations they admit:
 * computational basis states (which can be specified by an `int`),
 * pure states (which can be described by a 2^n-dimensional array or a (2, 2, ..., 2)-shaped tensor),
 * mixed states (which can be represented as a 2^n-by-2^n positive matrix with unit trace),
 * stabilizer states (which can be described by generators of their stabilizer group),
 * graph states (stabilizer states whose stabilizer generators can be described by a graph),
 * MPS states (which can be described by a simple tensor network),
 * PEPS states (which can be described by a little more complex tensor network),
 * ...

In various places in cirq we have the need to represent states in some of the sets above (e.g. mixed states in `cirq.DensityMatrixSimulator` and MPS states in `cirq.contrib.quimb.mps_simulator`).

A while back we discussed the question whether we should have a class to encapsulate a generic quantum state to hide the zoo behind a consistent interface. We decided not to do that and instead allow quantum states to be specified by simple primitive and numpy types (`int`, `np.ndarray`). This made sense at the time since most code manipulating quantum states was particular to the specific use-case and given a small subset of the zoo that we needed to support it was easy to infer the nature of the state based on the type and shape information. We also [defined](https://github.com/quantumlib/Cirq/pull/2376) a union type `cirq.STATE_VECTOR_LIKE` to describe the subset of the zoo we found relevant to us. This consisted of only four cases.

By now, we have added a class `cirq.ProductState` to `cirq.STATE_VECTOR_LIKE` and also added another union type `cirq.QUANTUM_STATE_LIKE` which encompasses `cirq.STATE_VECTOR_LIKE` and some additional options including another class for representing quantum states called `cirq.QuantumState`. We also have a mixin called `cirq.StateVectorMixin`...

In summary, we had considered a choice between a lightweight class-free approach of a simple union whose members were either primitive or maintained elsewhere (numpy) and a more heavy-weight approach with a single class to encapsulate the zoo. We ended up with the worst of both worlds :-) We now have two union types each with a class encapsulating a subset of the states. This means that on one hand code handling quantum states needs to deal with the diversity of the zoo and we lack a common place to extend quantum states with new functionality and on the other hand we have three classes to maintain.

**Describe your design idea/issue**

Perhaps, it is time to re-evaluate. One approach merges `cirq.ProductState`, `cirq.QuantumState` and `cirq.StateVectorMixin` and adds `cirq.state(...)` which can take as arguments any of the present members of the two union types. For example, we could say things like this

```python
cirq.state(0)                                   # |0>
cirq.state(np.diag([1, np.exp(1j*np.pi / 4)]))  # magic T state
cirq.state(np.eye(2) / 2, n_qubits=1)           # maximally mixed state of one qubit
cirq.state(np.eye(2) / np.sqrt(2), n_qubits=2)  # Bell state
cirq.state(stabilizers={cirq.Z})                # |0>
```

The factory `cirq.state` would continue to allow us to use simple ways to specify quantum states such as specifying computational basis states using an `int` while also providing us with a single place to add general state-related functionality. In particular, the new class would be a natural place for code that
* converts between state representations (e.g. producing density matrices of state vectors or stabilizer tableau from computational basis index etc),
* interprets a numpy array (together with optional hints) as a state (which can be done in a few different ways depending on Hilbert space dimension and qubit/qudit count),
* checks whether a state fulfills some basic condition (e.g. is pure, entangled, separable),
* generates human-readable notation (e.g. Dirac notation).

Filing to start discussion. Context: #3517.

### PR DESCRIPTION
**Pull request by daxfohl**
This PR unifies the quantum state types between `sim` and `qis` modules.

We create a `HasQuantumState` interface, from which both `QuantumState` and `ProductState` (used in qis), and `QuantumStateRepresentation` (the base class of quantum states used in simulators, including CliffordTableau, MPS, CHForm, SV/DM) inherit.

We then update the relevant `qis` method parameters to use `HasQuantumState` instead of `QuantumState`. This allows a broader range of quantum state representations to be analyzed in our `qis` utilities.

xref #4582

### CODE REVIEW

File: cirq-core/cirq/qis/measures_test.py [hunk 1]
```diff
@@ -164,6 +166,77 @@ def test_fidelity_product_states():
...
+        cirq.Z: 1,
+    }
+    circuits = [
+        cirq.testing.random_circuit(qubits=5, n_moments=30, op_density=1, gate_domain=gate_domain)
+        for _ in range(2)
+    ]
+
+    def final_state(sim: cirq.SimulatorBase):
+        return [
+            sim.simulate(c)._final_step_result_typed._sim_state.create_merged_state()._state
```
Comment (daxfohl) [inline cirq-core/cirq/qis/measures_test.py:190]: @95-martin-orion One thing to note here is that all of these states, regardless of the simulator, are found in the same (obscure) place. I think we should make this public after renaming ActOnArgs to SimulationState.

File: cirq-core/cirq/qis/states.py [hunk 2]
```diff
@@ -115,38 +166,14 @@ def state_vector(self) -> Optional[np.ndarray]:
...
-
-        A state tensor stores the amplitudes of a pure state as an array with
-        shape equal to the qid shape of the state.
-        If the state is a density matrix, this method returns None.
-        """
-        if self._is_density_matrix():
-            return None
-        return np.reshape(self.data, self.qid_shape)
-
     def density_matrix(self) -> np.ndarray:
```
Comment (95-martin-orion) [inline cirq-core/cirq/qis/states.py:169]: Keeping this here seems wrong - density matrices should override it, and everything else can call the parent method.

File: cirq-core/cirq/value/product_state.py [hunk 3]
```diff
@@ -63,6 +64,9 @@ def __init__(self, states=None):
             # coverage: ignore
             states = dict()
 
+        n = len(states)
+        object.__setattr__(self, '_qid_shape', (2,) * n)
+        object.__setattr__(self, '_dim', 2**n)
```
Comment (95-martin-orion) [inline cirq-core/cirq/value/product_state.py:69]: Does `super().__init__()` violate the `frozen=True` contract?

File: cirq-core/cirq/qis/states.py [hunk 4]
```diff
@@ -46,14 +47,70 @@
...
     # quantum state object
-    'cirq.QuantumState',
+    'HasQuantumState',
 ]
 document(QUANTUM_STATE_LIKE, """An object representing a quantum state.""")  # type: ignore
 
 
-class QuantumState:
-    """A quantum state.
+class HasQuantumState(abc.ABC):
```
Comment (95-martin-orion) [inline cirq-core/cirq/qis/states.py:55]: Mixed feelings on this name - I'd expect "Has..." to be a method, intuitively. Is `AbstractQuantumState` already taken?

**Review by 95-martin-orion — CHANGES_REQUESTED** [review]
What all does this actually buy us? I assume `qis` makes use of `QUANTUM_STATE_LIKE` in several places, and that's how this gets distributed? OTherwise it seems like only `von_neumann_entropy` is benefiting from this.

**Comment by daxfohl** [comment]
@95-martin-orion Perhaps it's not useful. I'm mostly looking at #4582, which seems to be asking for some unification between various types of quantum states allowing them to be used in qis independently, and then a `state` function that can create the best type. This PR provides the first part of that, and makes the second part much easier.

Admittedly when I started this I assumed `qis` was much richer than it is and that this would be a valuable addition. But, yes currently this only applies to `von_neumann_entropy` and `fidelity`, which is rather disappointing. Though I'm pretty happy with how the new unit tests look (though whether anyone would actually use the functionality like that, IDK)

I still think it's a nice unification for the sake of unification, but we should probably sync with @viathor and see if this is what he had in mind for #4582 next cynq. I'll re-mark this as draft until then.

File: cirq-core/cirq/value/product_state.py (hunk 3 above)
Comment (daxfohl) [reply cirq-core/cirq/value/product_state.py:69]: Yes, I considered changing the superclass initializer to the above. That allowed use of `super.__init__()` here, but looked ugly and created "_qid_shape is not defined in init" warnings. This seemed like the best alternative.

**Comment by daxfohl** [comment]
@95-martin-orion The other consideration is I remember a while ago there was a proposal to adopt some existing qis library. That may actually make sense here. The only real core thing in qis seems to be the `to_valid_stste_vector/density_matrix` functions that we could just move to linalg. (I believe the implementation of those could be simplified a bit too. Or, it's likely that quimb or whatever has equivalents.) But beyond that, the qis module seems pretty sparse and unmaintained, and not really integrated with anything else.

**Pull request closed without merging**